ast/src/parsed/mod.rs

```diff
@@ -320,6 +338,32 @@ impl<R> Children<Expression<R>> for EnumVariant<Expression<R>> {
     }
 }
 
+#[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Clone, Serialize, Deserialize, JsonSchema)]
+pub struct TraitImplementation<Expr> {
+    pub name: String,
+    pub type_scheme: Option<TraitScheme<Expr>>,
+    pub functions: Vec<NamedExpression<Expr>>,
+}
+
+impl<R> Children<Expression<R>> for TraitImplementation<Expression<R>> {
+    fn children(&self) -> Box<dyn Iterator<Item = &Expression<R>> + '_> {
+        Box::new(self.functions.iter().flat_map(|m| m.body.children()))
```

Comment:
`type_scheme` has an `Expr` parameter as well, don't forget that!

Reply:
By the way, this might get a bit complicated later on, because in expression processor change the types from `Type<Expression>` to `Type<u64>`, so we probably need a new type for trait impls after that.

The alternative is to parse it as `Type<u64>` right away - which is probably good enough since we cannot handle statically-sized array types in the type checker any way.

I did the same here: https://github.com/powdr-labs/powdr/pull/1545/files#diff-786a78af64899fc3240c68b7f68cb48a1f9e5772564adbe2cf3d53e08720a902R737

Reply:
Yes, in the giant PR on Traits I had to create both versions as you mentione here. I definitely think it's better to improve the types in the grammar rather than duplicate TraitImplementation so I'm basing this one on the types in your PR to avoid duplication.